Fixes #10275 -Use secure cookies when SSL

shlomizadok opened the pull request:
In production, we must enforce ssl, so cookies will be secured :)
Tried on a local foreman (Satellite machine):
![secure](https://cloud.githubusercontent.com/assets/433583/7368299/5cee4b22-edaf-11e4-9b49-c810620c0f3a.png)

This should be backported to 1.7 and 1.8

domcleal commented:
Does this affect non-HTTPS usage?  What if require_ssl's disabled?

Also the other cookies mentioned in my comment to the ticket should be fixed.

ohadlevy commented on config/environments/production.rb:
```diff
@@ -131,7 +131,7 @@
 
   # Force all access to the app over SSL, use Strict-Transport-Security, and use secure cookies.
-  # config.force_ssl = true
+  config.force_ssl = true
```
won't this break provisioning? (which needs http)

domcleal commented:
Setting back to WoC for answers to all of the above questions.

shlomizadok commented:
Added secure params to $.cookie on timezone and selected hosts cookies.
When in ssl:
![foreman_dev_secure](https://cloud.githubusercontent.com/assets/433583/7369751/196f0180-edbd-11e4-8a83-cd948baa686a.png)

When without ssl:

![foreman_dev_insecure](https://cloud.githubusercontent.com/assets/433583/7369758/22ab5d0c-edbd-11e4-9c64-b767c6493e56.png)

One thing I am not sure about is provisioning. Any thoughts @domcleal ?

domcleal commented:
> One thing I am not sure about is provisioning. Any thoughts @domcleal ?

I don't follow, what aren't you sure about?

ohadlevy commented:
@shlomizadok i guess you can use force_ssl:
http://api.rubyonrails.org/classes/ActionController/ForceSSL/ClassMethods.html

but, imho you would need to look at:
1. disabling existin ssl redication code in foreman
2. skipping it for unattended controller conditionally
3. check that the smart proxy auth code still works as expected

anything else @domcleal ?

ohadlevy commented on config/environments/production.rb:
```diff
@@ -131,7 +131,7 @@
 
   # Force all access to the app over SSL, use Strict-Transport-Security, and use secure cookies.
-  # config.force_ssl = true
+  config.force_ssl = true if SETTINGS[:require_ssl]
```
``` ruby
config.force_ssl =  !!SETTINGS[:require_ssl]
```

dLobatog replied:
That seems a bit dangerous if the file is misconfigured, `'false'` gets evaluated to true.

dLobatog commented on app/assets/javascripts/application.js:
```diff
@@ -427,3 +427,10 @@ function setPowerState(item, status){
 }
+
+function isSecurePage() {
+  if (location.protocol === 'https:')
```
Just `return location.protocol === 'https:'` suffices, but if that's all the method does I guess it can just be put above, no need to jump 300 lines to know what line 114 is doing.

shlomizadok replied:
lol... it is also used by `$.cookie($.cookieName, JSON.stringify($.foremanSelectedHosts), { secure: isSecurePage() });` which jumps to another file :bowtie: 
But I understand what you're saying, will refactor

shlomizadok commented on test/functional/hosts_controller_test.rb:
```diff
@@ -578,6 +578,10 @@ def test_unset_manage
     Setting[:require_ssl_smart_proxies] = true
     SETTINGS[:require_ssl] = true
+    # Adding https and certificates or else the request will be redirected
+    @request.env['HTTPS'] = 'on'
```
Since `SETTINGS[:require_ssl] = true`, I have to simulate an https call, or else I'll be redirected to https (and the response will be 301, and not 403 as expected)

domcleal replied:
That means you're no longer testing the scenario described on line 576?  Instead it seems to test the SSL verification step over HTTPS.

That probably points to this PR needing to add something to disable the redirect in the Foreman::Controller::SmartProxyAuth concern, which currently disables the existing SSL redirect and checking require_ssl itself.

domcleal commented:
Does this obsolete our own require_ssl filter?

shlomizadok commented:
basically it turned our own `require_ssl` filter to redirect to https if ssl is required

domcleal commented on app/controllers/application_controller.rb:
```diff
@@ -61,13 +62,15 @@ def deny_access
+    SETTINGS[:require_ssl]
+  end
+
   def require_ssl
```
How is this used?

shlomizadok replied:
It is used in https://github.com/theforeman/foreman/pull/2328/files#diff-55c5b7aecfb519d0e4880eaf2788eb6eR4
which is like a before_filter

domcleal replied:
That's `require_ssl?` (trailing question mark), while my question is about this specific line, an existing filter called `require_ssl`.

shlomizadok replied:
atm, `require_ssl` serves as redirect to 'https' when the request is  is from 'http'.
The rest has moved to `require_ssl?`

domcleal replied:
Isn't that the same as `force_ssl`?  What I'm getting at, is why we now have two systems for HTTP to HTTPS direction.

domcleal commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -4,10 +4,13 @@
   module ClassMethods
     def add_smart_proxy_filters(actions, options = {})
       skip_before_filter :require_login, :only => actions
-      skip_before_filter :require_ssl, :only => actions
```
the `require_ssl?` method should only skip the SSL requirement on `actions`, not controller-wide

domcleal commented on app/controllers/unattended_controller.rb:
```diff
@@ -1,6 +1,6 @@
 class UnattendedController < ApplicationController
   layout false
-
+  force_ssl :if => :require_ssl?
```
Any need to specify this again, since it's in appcontroller?

domcleal commented on app/controllers/unattended_controller.rb:
```diff
@@ -67,6 +67,12 @@ def pxe_config
 
+  protected
+
+  def require_ssl?
```
This also isn't equivalent, the previous code only skipped the SSL filter on certain actions and when there's no spoof or hostname parameter.

shlomizadok replied:
I honestly do not know it worked so far...
It seems that all of the actions require non-ssl connection

domcleal replied:
Sure, it "works", but previously it would still require SSL if the user was accessing the spoof view (from host#show > Templates > Review).  This change will allow that to happen over a non-SSL connection, which shouldn't be the case.

domcleal commented on config/environments/production.rb:
```diff
@@ -131,7 +131,7 @@
 
   # Force all access to the app over SSL, use Strict-Transport-Security, and use secure cookies.
-  # config.force_ssl = true
+  config.force_ssl = !!SETTINGS[:require_ssl]
```
It might be best to add this to dev/test too for consistency.

shlomizadok replied:
maybe remove from production.rb and add to config/application.rb ?

domcleal replied:
Yeah, that'd work.  At the moment, require_ssl is used in all environments so we should probably be consistent.  I can't think of a reason to not have it in development, since it's configurable and defaults to off.

domcleal commented on test/functional/hosts_controller_test.rb:
```diff
@@ -578,7 +578,6 @@ def test_unset_manage
     Setting[:restrict_registered_smart_proxies] = true
     Setting[:require_ssl_smart_proxies] = true
     SETTINGS[:require_ssl] = true
-
```
unrelated, restore please

domcleal commented:
Please just add a comment to say when you've updated and what's changed, it's hard to realise otherwise.

shlomizadok commented:
@domcleal - thanks for reviewing and the patience.
added a `skip_before_filter :force_ssl, :only => actions` to smart_proxy_auth,
changed `require_ssl?` on unattended controller, to reflect when to skip ssl

domcleal commented:
There's a test failure, can you take a look please?

<pre>
<403> expected but was
<301>. (test_0039_when "require_ssl_smart_proxies" and "require_ssl" are true, HTTP requests should not be able to get externalNodes)
/var/lib/workspace/workspace/test_develop_pr_core@4/database/mysql/ruby/1.9.3/test/functional/hosts_controller_test.rb:584
</pre>

shlomizadok commented:
yeah, fixing this now

On Thu, May 7, 2015 at 4:07 PM Dominic Cleal notifications@github.com
wrote:

> There's a test failure, can you take a look please?
> 
> <403> expected but was
> <301>. (test_0039_when "require_ssl_smart_proxies" and "require_ssl" are true, HTTP requests should not be able to get externalNodes)
> /var/lib/workspace/workspace/test_develop_pr_core@4/database/mysql/ruby/1.9.3/test/functional/hosts_controller_test.rb:584
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/theforeman/foreman/pull/2328#issuecomment-99858395.

shlomizadok commented:
@domcleal, Added `force_ssl` to api + added a lovely method to return false on smart_proxy_auth actions.

domcleal commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -7,16 +7,25 @@ module Foreman::Controller::SmartProxyAuth
       before_filter(:only => actions) { require_smart_proxy_or_login(options[:features]) }
       attr_reader :detected_proxy
+      define_method("require_ssl_with_dont?") do
+        send("require_ssl?") unless actions
```
`unless actions` will always be false.. shouldn't that check if `actions.include?(the current action)`?

domcleal commented on app/controllers/concerns/foreman/controller/smart_proxy_auth.rb:
```diff
@@ -7,16 +7,25 @@ module Foreman::Controller::SmartProxyAuth
 
+  protected
+
+  def require_ssl_with_dont?
```
is this required?  I can't see how it'd be called, since you define the same method in the add_smart_proxy_filters method, and only call alias_method_chain at the same time.

shlomizadok commented:
@domcleal, not sure it's the right approach (though I really hope so... ) so I separated that to a different commit (will squash is ack by you)

domcleal commented:
@shlomizadok I'm afraid it isn't, as you're defining a controller-method that is always going to return false, it isn't defined per action.  I think you were a bit closer with the previous method, but you needed to check if `actions.include?(action)` inside the method when working out what to return (remember, it's being run on each request).

domcleal commented:
You were very close I think, but the logic calling the "without" method was inverted I reckon.  You need to call the original method (that returns true) on all other actions, while currently it was called if it was the skipped action.

The latest commit on https://github.com/domcleal/foreman/commits/10275-forcessl-smartproxyauth fixes this and tidies the logic a little.  Those functional tests pass now here, haven't run the full suite.

domcleal commented:
When trying to POST to /api/v2/hosts/facts or access /unattended/provision over HTTP, I'm getting redirected immediately to HTTPS.  These routes should be excluded, since you overrode `require_ssl?` but it seems not:

<pre>
[root@foreman result]# curl -k -i -X POST -d '{}' "http://foreman.example.com/api/v2/hosts/facts"                                       
HTTP/1.1 301 Moved Permanently
Date: Fri, 08 May 2015 13:34:39 GMT
Server: Apache/2.4.6 (CentOS)
Strict-Transport-Security: max-age=31536000
X-Rack-Cache: invalidate, pass
X-Powered-By: Phusion Passenger 4.0.18
Location: https://foreman.example.com/api/v2/hosts/facts
Status: 301 Moved Permanently
Content-Length: 0
Connection: close
Content-Type: text/html; charset=UTF-8
[root@foreman result]# curl -k -i "http://foreman.example.com/unattended/provision"                                                     
HTTP/1.1 301 Moved Permanently
Date: Fri, 08 May 2015 13:38:39 GMT
Server: Apache/2.4.6 (CentOS)
Strict-Transport-Security: max-age=31536000
X-Rack-Cache: miss
X-Powered-By: Phusion Passenger 4.0.18
Location: https://foreman.example.com/unattended/provision
Status: 301 Moved Permanently
Content-Length: 0
Connection: close
Content-Type: text/html; charset=UTF-8
</pre>

I think config.force_ssl option forces _all_ traffic over HTTPS without it even getting to the controller's force_ssl filter and permitting a per-action behaviour.

domcleal commented:
The redirect behaviour appears to come from rack/ssl, loaded here: https://github.com/rails/rails/blob/v3.2.20/railties/lib/rails/application.rb#L251-L254

Is configuring the session options enough instead?  https://github.com/rails/rails/blob/v3.2.20/railties/lib/rails/application.rb#L282-L284

shlomizadok commented:
@domcleal - removed `config.force_ssl` option entirely, as we let the controllers decide on when to enforce ssl (because force_ssl on application.rb / <env> will force it anyhow). Hope this helps.

domcleal commented:
@shlomizadok okay, but now we're back to the issue you're originally aiming to fix - the session cookie has no secure flag.  See the link I mentioned in my last reply, about setting the session options - is that an alternative way?

(Currently this PR only fixes two non-session cookies, and simply refactors our redirects.)

shlomizadok commented:
@domcleal - another try at application.rb to set cookie to secure.

domcleal commented on config/application.rb:
```diff
@@ -146,6 +146,12 @@ class Application < Rails::Application
     end
+
+    # Secure cookies if the connection is via SSL
+    if SETTINGS[:require_ssl] == true
```
`if !!SETTINGS[:require_ssl]` for consistency

shlomizadok replied:
Chose the `true` option :)

domcleal replied:
My only nitpick is that it's different to https://github.com/theforeman/foreman/pull/2328/files#diff-b4f92b4e361ffab4fbb91c68eb3f77bbR69 in behaviour.

domcleal commented on config/application.rb:
```diff
@@ -146,6 +146,12 @@ class Application < Rails::Application
+
+    # Secure cookies if the connection is via SSL
+    if SETTINGS[:require_ssl] == true
+      config.session_options[:secure] = !!SETTINGS[:require_ssl]
```
or `true`

domcleal commented:
Alright, looks good and tests well, thanks @shlomizadok.

Since this has evolved into two different things, would you mind moving the force_ssl changes into a separate PR against ticket http://projects.theforeman.org/issues/10471?  We'll merge both, but that will probably go to 1.9.0 and this to our stable release(s) as we don't need it to fix the CVE.

Thanks!

shlomizadok committed:
fixes #10275 - Add secure cookie when in ssl

shlomizadok commented:
Done. #2373 is the second part

domcleal commented:
Merged as 0b03b9b, thanks @shlomizadok!
